docs/performance-metrics.md

```diff
@@ -0,0 +1,98 @@
+# OpenShift ETCD Performance Metrics
+
+We often get complaints about ETCD performance being bad, this document should help track down the culprit and mitigate the issue at hand using Prometheus Metrics.
+
+## Common Bottlenecks
+
+### Memory
+
+ETCD uses bbolt, which stores everything as a single memory mapped file. In OpenShift this is up to 8 GiB big, this file is organized with a B+Tree and chunked into discrete 4k pages which can be flagged in-use or not. On top of that, several in-memory caches/indices on both ETCD and API server are being used to facilitate fast multi-version access and label/field selection respectively.
+
+Due to the indices and caches employed, it is fairly easy to exhaust memory on the control plane even though the actual used size of the etcd database file is tiny. One particular scenario where is this common is when listing all pods in a large cluster or listing many large configmaps/secrets.
+
+Check these metrics:
+
+> process_resident_memory_bytes{job="etcd"}
+> 
+> process_resident_memory_bytes{job="apiserver"}
+
+It is perfectly normal for etcd to use a multiple of its "in use DB size" as resident memory.
+
+Try to look through the audit log and find service accounts that might be listing pods or configmaps across namespaces.
+Another frequent offender are events from chatty operators, which can be found in the audit log or directly from [etcd using KCS6970015](https://access.redhat.com/solutions/6970015).
+
+#### Defrag and Compaction
+
+One way to reduce memory usage is to defrag bbolt and compact the database to the latest revision. It is important to run the defrag after the compaction, so one would execute:
+
+> etctctl compact $(etctctl get / -w json | jq .header.revision)
+> 
+> etctctl defrag
+
+In more recent OpenShift versions this is done automatically -  compacting from the API server and defrag from the etcd-operator.
+
+Sometimes etcd is quite reluctant to free memory, due to the laziness of the Golang GC. So it is perfectly normal for the resident memory to stay up until memory pressure forces it to go down again.
+
```

Comment:
Also interesting to see the automatic defrag operation in action if you just plugin `etcd_mvcc_db_total_size_in_use_in_bytes` and see the sawtooth for the periodic defrag.

<img width="1611" alt="Screen Shot 2022-08-30 at 3 35 35 PM" src="https://user-images.githubusercontent.com/6660644/187555537-90f749ed-1d90-4ac4-9256-898fe0cdb6af.png">


Although it's probably easier to just look at the logs to see if the defrag call is succeeding rather than trying to infer it from this metric so not super useful.

Reply:
yep, I think the issue is that this exact graph is shown in the console metric dashboard for etcd. I guess lots of users freak out over this not reducing immediately.

I'll add a quick statement around checking the operator logs.